src/Utilities/ToolLocationHelper.cs

```diff
@@ -1381,6 +1386,133 @@ public static string GetLatestSDKTargetPlatformVersion(string sdkIdentifier, str
+        /// <param name="subFolder">Optional. If specified, the returned list will contain [VSInstallDir][subFolder] paths</param>
+        /// <returns>Returns folders in VS installs in VS version descending order separated by ';'</returns>
+        public static string GetFoldersInVSInstallsAsString(string minVersionString = null, string maxVersionString = null, string subFolder = null)
+        {
+            string foldersString = string.Empty;
+
+            try
+            {
+                Version minVersion, maxVersion;
+
+                if (string.IsNullOrEmpty(minVersionString) || !Version.TryParse(minVersionString, out minVersion))
+                {
+                    minVersion = null;
+                }
+                if (string.IsNullOrEmpty(maxVersionString) || !Version.TryParse(maxVersionString, out maxVersion))
+                {
+                    maxVersion = null;
+                }
+
+                var folders = GetFoldersInVSInstalls(minVersion, maxVersion, subFolder);
+
+                if (folders.Count() > 0)
+                {
+                    foldersString = string.Join(";", folders);
+                }
+            }
+            catch(Exception e)
+            {
+                // this method will be used in vc props and we don't want to fail project load if it throws for some non critical reason.
+                if (ExceptionHandling.IsCriticalException(e))
+                {
+                    throw;
+                }
+            }
+
+            return foldersString;
+        }
+
+        /// <summary>
+        /// Finds first folder in the list which contains all given files. Returns an empty string if not found.
+        /// </summary>
+        /// <param name="possibleRoots">Root folders separated by ';'</param>
+        /// <param name="relativeFilePaths">Relative file paths to find under root folders, separated by ';'.</param>
+        /// <returns>The first root folder in the given list, which contains all files. Empty string if not found.</returns>
+        public static string FindRootFolderWhereAllFilesExist(string possibleRoots, string relativeFilePaths)
+        {
+            if (!string.IsNullOrEmpty(possibleRoots))
+            {
+                var roots = possibleRoots.Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
+                var files = relativeFilePaths.Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
+
+                bool allFilesFound;
+                foreach (var root in roots)
+                {
+                    allFilesFound = true;
+                    foreach (var file in files)
+                    {
+                        try
+                        {
+                            string fullPath = Path.Combine(root, file);
+
+                            if (!FileSystems.Default.FileExists(fullPath))
```

Comment:
This seems like a place where we'd want to use a cached evaluation context. But I don't know if it's plumbed into the right places.

(This can wait; it's an optimization)